File: data_structure/heap/src/HeapDemo.rb
# Ruby
# データ構造: ヒープ (Heap)

class HeapData
  def initialize(is_min_heap = true)
    @data = []
    @is_min_heap = is_min_heap
  end

  def _get_parent_idx(idx)
    (idx - 1) / 2
  end

  def _get_left_child_idx(idx)
    2 * idx + 1
  end

  def _get_right_child_idx(idx)
    2 * idx + 2
  end

  def _has_parent?(idx)
    _get_parent_idx(idx) >= 0
  end

  def _has_left_child?(idx)
    _get_left_child_idx(idx) < @data.length
  end

  def _has_right_child?(idx)
    _get_right_child_idx(idx) < @data.length
  end

  def _get_parent(idx)
    @data[_get_parent_idx(idx)]
  end

  def _get_left_child(idx)
    @data[_get_left_child_idx(idx)]
  end

  def _get_right_child(idx)
    @data[_get_right_child_idx(idx)]
  end

  def _swap(idx1, idx2)
    @data[idx1], @data[idx2] = @data[idx2], @data[idx1]
  end

  def _should_swap?(idx1, idx2)
    if @is_min_heap
      @data[idx1] > @data[idx2]
    else
      @data[idx1] < @data[idx2]
    end
  end

  def _heapify_down(idx)
    smallest_or_largest = idx

    if _has_left_child?(idx) && _should_swap?(_get_left_child_idx(idx), smallest_or_largest)
      smallest_or_largest = _get_left_child_idx(idx)
    end

    if _has_right_child?(idx) && _should_swap?(_get_right_child_idx(idx), smallest_or_largest)
      smallest_or_largest = _get_right_child_idx(idx)
    end

    if smallest_or_largest != idx
      _swap(idx, smallest_or_largest)
      _heapify_down(smallest_or_largest)
    end
  end

  def _heapify_up(idx)
    while _has_parent?(idx) && _should_swap?(_get_parent_idx(idx), idx)
      parent_idx = _get_parent_idx(idx)
      _swap(parent_idx, idx)
      idx = parent_idx
    end
  end

  def get
    @data
  end

  def get_index(item)
    begin
      @data.index(item)
    rescue
      puts "ERROR: #{item} は範囲外です"
      -1
    end
  end

  def get_value(index)
    if index >= 0 && index < @data.length
      @data[index]
    else
      puts "ERROR: #{index} は範囲外です"
      nil
    end
  end

  def heapify(array)
    @data = array.dup
    (@data.length / 2 - 1).downto(0) do |i|
      _heapify_down(i)
    end
    true
  end

  def push(value)
    @data << value
    _heapify_up(@data.length - 1)
    true
  end

  def pop
    return false if @data.empty?

    last_element = @data.pop

    if @data.length > 0
      @data[0] = last_element
      _heapify_down(0)
    end

    true
  end

  def peek
    return nil if @data.empty?
    @data[0]
  end

  def empty?
    @data.empty?
  end

  def size
    @data.length
  end

  def clear
    @data = []
    true
  end
end

def main
  puts "Heap TEST -----> start"

  puts "\nmin heap: new"
  min_heap = HeapData.new(true)
  puts "  現在のデータ: #{min_heap.get}"

  puts "\nmin heap: heapify"
  input = [4, 10, 3, 5, 1]
  puts "  入力値: #{input}"
  output = min_heap.heapify(input)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{min_heap.get}"

  puts "\nmin heap: push"
  input = 2
  puts "  入力値: #{input}"
  output = min_heap.push(input)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{min_heap.get}"

  puts "\nmin heap: push"
  input = 15
  puts "  入力値: #{input}"
  output = min_heap.push(input)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{min_heap.get}"

  puts "\nmin heap: peek"
  output = min_heap.peek
  puts "  出力値: #{output}"

  puts "\nmin heap: pop"
  output = min_heap.pop
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{min_heap.get}"

  puts "\nmin heap: pop"
  output = min_heap.pop
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{min_heap.get}"

  puts "\nmin heap: get_index"
  input = 3
  puts "  入力値: #{input}"
  output = min_heap.get_index(input)
  puts "  出力値: #{output}"

  puts "\nmin heap: get_index"
  input = 100
  puts "  入力値: #{input}"
  output = min_heap.get_index(input)
  puts "  出力値: #{output}"

  puts "\nmin heap: is_empty"
  output = min_heap.empty?
  puts "  出力値: #{output}"

  puts "\nmin heap: size"
  output = min_heap.size
  puts "  出力値: #{output}"

  puts "\nmin heap: clear"
  output = min_heap.clear
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{min_heap.get}"

  puts "\nmin heap: is_empty"
  output = min_heap.empty?
  puts "  出力値: #{output}"

  puts "\nmax heap: new"
  max_heap = HeapData.new(false)
  puts "  現在のデータ: #{max_heap.get}"

  puts "\nmax heap: heapify"
  input = [4, 10, 3, 5, 1]
  puts "  入力値: #{input}"
  output = max_heap.heapify(input)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{max_heap.get}"

  puts "\nmax heap: push"
  input = 12
  puts "  入力値: #{input}"
  output = max_heap.push(input)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{max_heap.get}"

  puts "\nmax heap: push"
  input = 0
  puts "  入力値: #{input}"
  output = max_heap.push(input)
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{max_heap.get}"

  puts "\nmax heap: peek"
  output = max_heap.peek
  puts "  出力値: #{output}"

  puts "\nmax heap: pop"
  output = max_heap.pop
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{max_heap.get}"

  puts "\nmax heap: pop"
  output = max_heap.pop
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{max_heap.get}"

  puts "\nmax heap: get_index"
  input = 5
  puts "  入力値: #{input}"
  output = max_heap.get_index(input)
  puts "  出力値: #{output}"

  puts "\nmax heap: get_index"
  input = -10
  puts "  入力値: #{input}"
  output = max_heap.get_index(input)
  puts "  出力値: #{output}"

  puts "\nmax heap: is_empty"
  output = max_heap.empty?
  puts "  出力値: #{output}"

  puts "\nmax heap: size"
  output = max_heap.size
  puts "  出力値: #{output}"

  puts "\nmax heap: clear"
  output = max_heap.clear
  puts "  出力値: #{output}"
  puts "  現在のデータ: #{max_heap.get}"

  puts "\nmax heap: is_empty"
  output = max_heap.empty?
  puts "  出力値: #{output}"

  puts "\nHeap TEST <----- end"
end

main if __FILE__ == $PROGRAM_NAME

File: data_structure/heap/src/HeapDemo.kt
// Kotlin
// データ構造: ヒープ (Heap)

class HeapData(private val isMinHeap: Boolean = true) {
    private val data = mutableListOf<Int>()

    private fun getParentIdx(idx: Int) = (idx - 1) / 2
    private fun getLeftChildIdx(idx: Int) = 2 * idx + 1
    private fun getRightChildIdx(idx: Int) = 2 * idx + 2

    private fun hasParent(idx: Int) = getParentIdx(idx) >= 0
    private fun hasLeftChild(idx: Int) = getLeftChildIdx(idx) < data.size
    private fun hasRightChild(idx: Int) = getRightChildIdx(idx) < data.size

    private fun getParent(idx: Int) = data[getParentIdx(idx)]
    private fun getLeftChild(idx: Int) = data[getLeftChildIdx(idx)]
    private fun getRightChild(idx: Int) = data[getRightChildIdx(idx)]

    private fun swap(idx1: Int, idx2: Int) {
        val temp = data[idx1]
        data[idx1] = data[idx2]
        data[idx2] = temp
    }

    private fun shouldSwap(idx1: Int, idx2: Int): Boolean {
        return if (isMinHeap) {
            data[idx1] > data[idx2]
        } else {
            data[idx1] < data[idx2]
        }
    }

    private fun heapifyDown(idx: Int) {
        var smallestOrLargest = idx

        if (hasLeftChild(idx) && shouldSwap(smallestOrLargest, getLeftChildIdx(idx))) {
            smallestOrLargest = getLeftChildIdx(idx)
        }

        if (hasRightChild(idx) && shouldSwap(smallestOrLargest, getRightChildIdx(idx))) {
            smallestOrLargest = getRightChildIdx(idx)
        }

        if (smallestOrLargest != idx) {
            swap(idx, smallestOrLargest)
            heapifyDown(smallestOrLargest)
        }
    }

    private fun heapifyUp(idx: Int) {
        var currentIdx = idx
        while (hasParent(currentIdx) && shouldSwap(getParentIdx(currentIdx), currentIdx)) {
            val parentIdx = getParentIdx(currentIdx)
            swap(parentIdx, currentIdx)
            currentIdx = parentIdx
        }
    }

    fun get(): List<Int> = data.toList()

    fun getIndex(item: Int): Int {
        val index = data.indexOf(item)
        return if (index != -1) index else {
            println("ERROR: $item は範囲外です")
            -1
        }
    }

    fun getValue(index: Int): Int? {
        return if (index in data.indices) {
            data[index]
        } else {
            println("ERROR: $index は範囲外です")
            null
        }
    }

    fun heapify(array: List<Int>): Boolean {
        data.clear()
        data.addAll(array)
        for (i in data.size / 2 - 1 downTo 0) {
            heapifyDown(i)
        }
        return true
    }

    fun push(value: Int): Boolean {
        data.add(value)
        heapifyUp(data.size - 1)
        return true
    }

    fun pop(): Boolean {
        if (data.isEmpty()) return false

        val lastElement = data.removeAt(data.lastIndex)
        if (data.isNotEmpty()) {
            data[0] = lastElement
            heapifyDown(0)
        }
        return true
    }

    fun peek(): Int? = if (data.isNotEmpty()) data[0] else null

    fun isEmpty(): Boolean = data.isEmpty()

    fun size(): Int = data.size

    fun clear(): Boolean {
        data.clear()
        return true
    }
}

fun main() {
    println("Heap TEST -----> start")

    println("\nmin heap: new")
    val minHeap = HeapData(isMinHeap = true)
    println("  現在のデータ: ${minHeap.get()}")

    println("\nmin heap: heapify")
    val input = listOf(4, 10, 3, 5, 1)
    println("  入力値: $input")
    val output = minHeap.heapify(input)
    println("  出力値: $output")
    println("  現在のデータ: ${minHeap.get()}")

    println("\nmin heap: push")
    val input1 = 2
    println("  入力値: $input1")
    val output1 = minHeap.push(input1)
    println("  出力値: $output1")
    println("  現在のデータ: ${minHeap.get()}")

    println("\nmin heap: push")
    val input2 = 15
    println("  入力値: $input2")
    val output2 = minHeap.push(input2)
    println("  出力値: $output2")
    println("  現在のデータ: ${minHeap.get()}")

    println("\nmin heap: peek")
    val output3 = minHeap.peek()
    println("  出力値: $output3")

    println("\nmin heap: pop")
    val output4 = minHeap.pop()
    println("  出力値: $output4")
    println("  現在のデータ: ${minHeap.get()}")

    println("\nmin heap: pop")
    val output5 = minHeap.pop()
    println("  出力値: $output5")
    println("  現在のデータ: ${minHeap.get()}")

    println("\nmin heap: get_index")
    val input3 = 3
    println("  入力値: $input3")
    val output6 = minHeap.getIndex(input3)
    println("  出力値: $output6")

    println("\nmin heap: get_index")
    val input4 = 100
    println("  入力値: $input4")
    val output7 = minHeap.getIndex(input4)
    println("  出力値: $output7")

    println("\nmin heap: is_empty")
    val output8 = minHeap.isEmpty()
    println("  出力値: $output8")

    println("\nmin heap: size")
    val output9 = minHeap.size()
    println("  出力値: $output9")

    println("\nmin heap: clear")
    val output10 = minHeap.clear()
    println("  出力値: $output10")
    println("  現在のデータ: ${minHeap.get()}")

    println("\nmin heap: is_empty")
    val output11 = minHeap.isEmpty()
    println("  出力値: $output11")

    println("\nmax heap: new")
    val maxHeap = HeapData(isMinHeap = false)
    println("  現在のデータ: ${maxHeap.get()}")

    println("\nmax heap: heapify")
    val input5 = listOf(4, 10, 3, 5, 1)
    println("  入力値: $input5")
    val output12 = maxHeap.heapify(input5)
    println("  出力値: $output12")
    println("  現在のデータ: ${maxHeap.get()}")

    println("\nmax heap: push")
    val input6 = 12
    println("  入力値: $input6")
    val output13 = maxHeap.push(input6)
    println("  出力値: $output13")
    println("  現在のデータ: ${maxHeap.get()}")

    println("\nmax heap: push")
    val input7 = 0
    println("  入力値: $input7")
    val output14 = maxHeap.push(input7)
    println("  出力値: $output14")
    println("  現在のデータ: ${maxHeap.get()}")

    println("\nmax heap: peek")
    val output15 = maxHeap.peek()
    println("  出力値: $output15")

    println("\nmax heap: pop")
    val output16 = maxHeap.pop()
    println("  出力値: $output16")
    println("  現在のデータ: ${maxHeap.get()}")

    println("\nmax heap: pop")
    val output17 = maxHeap.pop()
    println("  出力値: $output17")
    println("  現在のデータ: ${maxHeap.get()}")

    println("\nmax heap: get_index")
    val input8 = 5
    println("  入力値: $input8")
    val output18 = maxHeap.getIndex(input8)
    println("  出力値: $output18")

    println("\nmax heap: get_index")
    val input9 = -10
    println("  入力値: $input9")
    val output19 = maxHeap.getIndex(input9)
    println("  出力値: $output19")

    println("\nmax heap: is_empty")
    val output20 = maxHeap.isEmpty()
    println("  出力値: $output20")

    println("\nmax heap: size")
    val output21 = maxHeap.size()
    println("  出力値: $output21")

    println("\nmax heap: clear")
    val output22 = maxHeap.clear()
    println("  出力値: $output22")
    println("  現在のデータ: ${maxHeap.get()}")

    println("\nmax heap: is_empty")
    val output23 = maxHeap.isEmpty()
    println("  出力値: $output23")

    println("\nHeap TEST <----- end")
}

// Run the main function
fun runMainFunction() {
    main()
}
